platform/posix/transport/src/mbedtls_pkcs11_posix.c

```diff
@@ -0,0 +1,958 @@
+    {
+        /* Perform the TLS handshake. */
+        do
+        {
+            mbedtlsError = mbedtls_ssl_handshake( &( pMbedtlsPkcs11Context->context ) );
+        } while( ( mbedtlsError == MBEDTLS_ERR_SSL_WANT_READ ) ||
+                 ( mbedtlsError == MBEDTLS_ERR_SSL_WANT_WRITE ) );
+
+        if( ( mbedtlsError != 0 ) || ( mbedtls_ssl_get_verify_result( &( pMbedtlsPkcs11Context->context ) ) != 0U ) )
+        {
+            LogError( ( "Failed to perform TLS handshake: mbedTLSError= %s : %s.",
+                        mbedtlsHighLevelCodeOrDefault( mbedtlsError ),
+                        mbedtlsLowLevelCodeOrDefault( mbedtlsError ) ) );
+            returnStatus = MBEDTLS_PKCS11_HANDSHAKE_FAILED;
+        }
+    }
+
+    /* Clean up on failure. */
+    if( returnStatus != MBEDTLS_PKCS11_SUCCESS )
+    {
+        contextFree( pMbedtlsPkcs11Context );
+    }
+    else
+    {
+        LogInfo( ( "(Network connection %p) Connection to %s established.", pNetworkContext, pHostName ) );
+    }
+
+    return returnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+void Mbedtls_Pkcs11_Disconnect( NetworkContext_t * pNetworkContext )
+{
+    MbedtlsPkcs11Context_t * pMbedtlsPkcs11Context = NULL;
+    int tlsStatus = 0;
+
+    if( ( pNetworkContext != NULL ) && ( pNetworkContext->pParams != NULL ) )
+    {
+        pMbedtlsPkcs11Context = pNetworkContext->pParams;
+        /* Attempting to terminate TLS connection. */
+        tlsStatus = mbedtls_ssl_close_notify( &( pMbedtlsPkcs11Context->context ) );
+
+        /* Ignore the WANT_READ and WANT_WRITE return values. */
+        if( ( tlsStatus != MBEDTLS_ERR_SSL_WANT_READ ) &&
+            ( tlsStatus != MBEDTLS_ERR_SSL_WANT_WRITE ) )
+        {
+            if( tlsStatus == 0 )
+            {
+                LogInfo( ( "(Network connection %p) TLS close-notify sent.",
+                           pNetworkContext ) );
+            }
+            else
+            {
+                LogError( ( "(Network connection %p) Failed to send TLS close-notify: mbedTLSError= %s : %s.",
+                            ( void * ) pNetworkContext,
+                            mbedtlsHighLevelCodeOrDefault( tlsStatus ),
+                            mbedtlsLowLevelCodeOrDefault( tlsStatus ) ) );
+            }
+        }
+        else
+        {
+            /* WANT_READ and WANT_WRITE can be ignored. Logging for debugging purposes. */
+            LogInfo( ( "(Network connection %p) TLS close-notify sent; ",
+                       "received %s as the TLS status can be ignored for close-notify."
+                       ( tlsStatus == MBEDTLS_ERR_SSL_WANT_READ ) ? "WANT_READ" : "WANT_WRITE",
+                       pNetworkContext ) );
+        }
+
+        /* Free contexts. */
+        contextFree( pMbedtlsPkcs11Context );
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+int32_t Mbedtls_Pkcs11_Recv( NetworkContext_t * pNetworkContext,
```

Comment:
We recently made changes to the Plaintext and OpenSSL transport implementations to add non-blocking behavior by using `poll` because we suggest the behavior in our documentation to customers. Please update these recv/send functions to have similar behavior